### PR TITLE
Add decimal rounding

### PR DESCRIPTION
**Pull request by iamyulong**
- Added `ceiling`, `floor`, and `round` methods to `Decimal`
- Removed `BigDecimal`
- Fixed an issue in amount check

### CODE REVIEW

File: radix-engine/src/engine/process.rs [hunk 1]
```diff
@@ -978,6 +978,12 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
 
         match mint_params {
             MintParams::Fungible { amount } => {
+                // It takes `1,701,411,835` mint operations to reach `Decimal::MAX`,
+                // which will be impossible with metering.
```
Comment (talekhinezh) [inline radix-engine/src/engine/process.rs:982]: lol

File: radix-engine/tests/bucket.rs [hunk 2]
```diff
@@ -46,3 +52,67 @@ fn test_bucket_of_badges() {
     let receipt = executor.run(transaction).unwrap();
     assert!(receipt.result.is_ok());
 }
+
+#[test]
+fn test_take_with_invalid_granularity() {
+    let mut substate_store = InMemorySubstateStore::with_bootstrap();
```
Comment (talekhinezh) [inline radix-engine/tests/bucket.rs:58]: // Arrange :)

File: radix-engine/tests/bucket.rs [hunk 3]
```diff
@@ -46,3 +52,67 @@ fn test_bucket_of_badges() {
...
+        receipt.result,
+        Err(RuntimeError::BucketError(
+            ResourceContainerError::InvalidAmount(dec!("1.123"), 2)
+        ))
+    );
+}
+
+#[test]
+fn test_take_with_negative_amount() {
+    let mut substate_store = InMemorySubstateStore::with_bootstrap();
```
Comment (talekhinezh) [inline radix-engine/tests/bucket.rs:90]: // Arrange :)

File: radix-engine/src/errors.rs [hunk 4]
```diff
@@ -183,9 +183,12 @@ pub enum RuntimeError {
     /// Index out of bounds.
     IndexOutOfBounds { index: usize, max: usize },
 
-    /// Mint parameters are invalid
+    /// Mint parameters are invalid.
     InvalidMintParams,
 
+    /// Attempted to mint more than the max allowed.
+    MaxMintAmountExceeded,
```
Comment (talekhinezh) [inline radix-engine/src/errors.rs:190]: Missing unit test for this error

File: radix-engine/tests/bucket.rs (hunk 2 above)
Comment (iamyulong) [reply radix-engine/tests/bucket.rs:58]: Haha, done

File: radix-engine/src/errors.rs (hunk 4 above)
Comment (iamyulong) [reply radix-engine/src/errors.rs:235]: Will add one

**Pull request merged**